src/features/device-state/routes/state-v3.ts

```diff
@@ -0,0 +1,412 @@
+				is_host?: boolean;
+				releases?: {
+					[uuid: string]: {
+						/**
+						 * @deprecated to be removed in state v4
+						 */
+						id: number;
+						services?: {
+							[name: string]: {
+								/**
+								 * @deprecated to be removed in state v4
+								 */
+								id: number;
+								/**
+								 * @deprecated to be removed in state v4
+								 */
+								image_id: number;
+								image: string;
+								/**
+								 * Defaults to true if undefined
+								 */
+								running?: boolean;
+								environment: {
+									[varName: string]: string;
+								};
+								labels: {
+									[labelName: string]: string;
+								};
+								contract?: AnyObject;
+								composition?: ServiceComposition;
+							};
+						};
+						volumes?: AnyObject;
+						networks?: AnyObject;
+					};
+				};
+			};
+		};
+	};
+};
+
+function buildAppFromRelease(
+	device: AnyObject,
+	application: AnyObject,
+	release: AnyObject,
+	config: Dictionary<string>,
+): NonNullable<LocalStateApp['releases']> {
+	let composition: AnyObject = {};
+	const services: NonNullable<LocalStateApp['releases']>[string]['services'] =
+		{};
+
+	// Parse the composition to forward values to the device
+	if (_.isObject(release.composition)) {
+		composition = release.composition;
+	} else {
+		try {
+			composition = JSON.parse(release.composition);
+		} catch {
+			composition = {};
+		}
+	}
+
+	(release.contains__image as AnyObject[]).forEach((ipr) => {
+		// extract the per-image information
+		const image = ipr.image[0];
+
+		const si = serviceInstallFromImage(device, image);
+		if (si == null) {
+			throw new Error(
+				`Could not find service install for device: '${
+					application.uuid
+				}', image: '${image?.id}', service: '${JSON.stringify(
+					image?.is_a_build_of__service,
+				)}', service installs: '${JSON.stringify(device.service_install)}'`,
+			);
+		}
+		const svc = si.service[0];
+
+		const environment: Dictionary<string> = {};
+		varListInsert(ipr.image_environment_variable, environment);
+		varListInsert(application.application_environment_variable, environment);
+		varListInsert(svc.service_environment_variable, environment);
+		varListInsert(device.device_environment_variable, environment);
+		varListInsert(si.device_service_environment_variable, environment);
+
+		const labels: Dictionary<string> = {};
+		[...ipr.image_label, ...svc.service_label].forEach(
+			({ label_name, value }: { label_name: string; value: string }) => {
+				labels[label_name] = value;
+			},
+		);
+
+		_.each(ConfigurationVarsToLabels, (labelName, confName) => {
+			if (confName in config && !(labelName in labels)) {
+				labels[labelName] = config[confName];
+			}
+		});
+
+		const imgRegistry =
+			image.is_stored_at__image_location +
+			(image.content_hash != null ? `@${image.content_hash}` : '');
+
+		services[svc.service_name] = {
+			id: svc.id,
+			image_id: image.id,
+			image: formatImageLocation(imgRegistry),
+			environment,
```

Comment:
The way this is setup, the environment and labels are sent twice as part of the target state. See the extract below for a test service. I am making sure the supervisor merges them correctly to use the right value, but it might be worth merging them here to reduce the payload size

```
"hello": {
    "id": 5,
    "image_id": 37,
    "image": "registry2.09e8bcd18440f7403fb526350f0350dc.bob.local/v2/c293641a12f4a18c0936e3c4d81debbb@sha256:56405843eda8dc5d6917430214e393825604f1de4927d59c5733ae777a00fbc9",
    "environment": {
      "MYVAR": "goodbye"
    },
    "labels": {},
    "composition": {
      "image": "my-custom-image",
      "environment": {
        "MYVAR": "hello"
      }
    }
  }
```

Reply:
@pipex the service level `environment` key includes image_environment_variable/application_environment_variable/service_environment_variable/device_environment_variable/device_service_environment_variable whereas the composition one only includes what is in the docker-compose.yml - I think the composition ones get brought into the api as image environment variables and so would show in both but they are definitely not the definitive list

Reply:
I don't understand what you are saying. I don't think that creates a problem as we know the priority, right? 

What I'm thinking is doing something like 

```
service[name]: {
    environment: {...composition.environment, ...environment}
}
``` 

which I'm [doing on the supervisor](https://github.com/balena-os/balena-supervisor/pull/1765/files#diff-c436eb4642e9d9fe975f53422f677d3b3c483de880eec12a003d401d67884f54R153-R154), but doing it on the API side would prevent sending extra data.

Reply:
Everything in the composition is already sent in the upper level environment @pipex with the correct priority. The reason for that is because the variables in the composition get added to the odata model on build. However it is then also possible to remove/update those entries from the model after the fact in which case we probably don't want to use what was initially set in the composition as they were explicitly changed.

Reply:
I think I'm missing the point as well :confused: 
Why does the supervisor needs to know both the effective env var values, as well as the image env vars from the docker-compose.yml separately? What is the use case that this enables for the supervisor?

Reply:
There's 2 separate things
1. the supervisor doesn't need to merge because the api already handles that
2. does the supervisor need the fully intact composition as was used to build? because if not there's probably a bunch of sections we can delete but I thought it was used for things? maybe contracts based on the composition? maybe livepush stuff? etc (I'm not really sure but if the supervisor doesn't actually care about having an intact composition we could probably delete a bunch)

Reply:
> the supervisor doesn't need to merge because the api already handles that

Got it, so we can only use the top-level composition

> does the supervisor need the fully intact composition as was used to build?

It does use the whole composition, except where it overrides the values with the cloud state, i.e. environment and labels. 

I see the point though. In local mode, it would probably make sense to put the environment variables in the composition as there are no device or fleet level overrides. So from a modeling perspective, the target state I shared above is more correct, although is more wasteful.

If we want to reduce the waste that might mean we need to tradeoff modeling correctness (which imho is reasonable)
- Move the environment/labels to inside the composition
- Remove the environment/labels from the composition since they are already mirrored in the top level property

I lean more towards the second option, which still allows to provide the composition properties if we ever need to